course/src/isometry.cpp

```diff
@@ -10,9 +10,135 @@
 namespace ekumen {
 namespace math {
 
-/*
- * your code here
- */
+namespace {
+
+Vector3 translation_vector_operation(const Matrix3& matrix,
+                                     const Vector3& first_vector,
+                                     const Vector3& second_vector) {
+  return matrix * first_vector + second_vector;
+}
+
+}  // namespace
+
+
+Isometry Isometry::fromTranslation(const Vector3& r_vector) {
+  return Isometry{r_vector, Matrix3::kIdentity};
+}
+
+Vector3 Isometry::translation() const {
+  return translation_vector;
+}
+
+Vector3 Isometry::transform(const Vector3& r_vector) const {
+  return (*this) * r_vector;
+}
+
+Isometry Isometry::inverse() const {
+  const double det = rotation_matrix.det();
+  if (det == 0) {
+    throw std::invalid_argument("Isometry doesn't have an inverse");
+  }
+  const Matrix3 transpose_matrix{rotation_matrix.transpose()};
+  const double c_00 = transpose_matrix[1][1] * transpose_matrix[2][2] -
+      transpose_matrix[1][2] * transpose_matrix[2][1];
+  const double c_01 = transpose_matrix[1][0] * transpose_matrix[2][2] -
+      transpose_matrix[1][2] * transpose_matrix[2][0];
+  const double c_02 = transpose_matrix[1][0] * transpose_matrix[2][1] -
+      transpose_matrix[1][1] * transpose_matrix[2][0];
+  const double c_10 = transpose_matrix[0][1] * transpose_matrix[2][2] -
+      transpose_matrix[0][2] * transpose_matrix[2][1];
+  const double c_11 = transpose_matrix[0][0] * transpose_matrix[2][2] -
+      transpose_matrix[0][2] * transpose_matrix[2][0];
+  const double c_12 = transpose_matrix[0][0] * transpose_matrix[1][2] -
+      transpose_matrix[0][2] * transpose_matrix[1][0];
+  const double c_20 = transpose_matrix[0][1] * transpose_matrix[1][2] -
+      transpose_matrix[0][2] * transpose_matrix[1][1];
+  const double c_21 = transpose_matrix[0][0] * transpose_matrix[1][2] -
+      transpose_matrix[0][2] * transpose_matrix[1][0];
+  const double c_22 = transpose_matrix[0][0] * transpose_matrix[1][1] -
+      transpose_matrix[0][1] * transpose_matrix[1][0];
+
+  Matrix3 aux{c_00, c_01, c_02, c_10, c_11, c_12, c_20, c_21, c_22};
+  aux = (aux * Matrix3{
+    1., -1., 1., -1., 1., -1., 1., -1., 1.});
+
+  aux /= det;
+  return Isometry{aux * translation_vector * -1, aux};
+}
+
+Matrix3 Isometry::rotation() const {
+  return rotation_matrix;
+}
+
+Isometry Isometry::compose(const Isometry& r_isometry) const {
+  Isometry aux = (*this) * r_isometry;
+  return aux;
```

Comment:
It's better if you return the result of the operation directly and avoid the intermediate variable: `return *this * r_isometry;`

Reply:
Addressed.